_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,256 @@
+    Stage       string       `gorm:"size:40;default:'';not null" json:"stage"`
+    Params      []BuildParam `gorm:"foreignKey:BuildID" json:"params"`
+    IsInvalid   bool         `gorm:"not null;default:false" json:"isInvalid"` 
+    // added 
+    TestResultSummaryCount uint `gorm:"not null" json:"testResultSummaryCount"`
+}
+// new
+type TestResultSummary struct {
+    ArtifactID  uint	  `gorm:"not null;index:testresultsummary_idx_artifact_id" json:"artifactId"`
+    Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    BuildID     uint      `gorm:"not null;index:testresultsummary_idx_build_id" json:"buildId"`
+    Build       *Build    `gorm:"foreignKey:BuildID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    RunCount    uint	  `gorm:"not null" json:"runCount"`
+    SkipCount   uint	  `gorm:"not null" json:"skipCount"`
+    FailCount   uint	  `gorm:"not null" json:"failCount"`
+    PassCount   uint	  `gorm:"not null" json:"passCount"`
+}
+// new
+type TestResult struct {
+    ArtifactID  uint 	  `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
+    Artifact    *Artifact `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
+    Name        string	  `gorm:"not null;" json:"name"`
+    Ran         string 	  `gorm:"not null;" json:"ran"`
+    Passed      string 	  `gorm:"not null;" json:"passed"`
+    StartedOn   *time.Time `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
+    CompletedOn *time.Time `gorm:"nullable;default:NULL;" json:"finishedOn" format:"date-time"`
+}
+```
+
+```go
+// artifact.go
+// new
+type File struct {
+    name string
+    fileName string
+    data []bytes
+}
+// modified
+func (m ArtifactModule) postBuildArtifactHandler(c *gin.Context) {
+    files := parseMultipartFormData(c)
+    buildId := ginutil.ParseParamUint(c, "buildid")
+    
+    for _, file := range files {
+    	storeArtifactInDB(file, buildID)
+    	if strings.HasSuffix(file.fileName, ".trx") {
+    	    parseTRXAndStoreInDB(file, buildID, artifact.ArtifactID)
+    	}
+    }
+}
+// new, /build/{buildid}/artifact/{artifactid}/test-results
+func (m ArtifactModule) getBuildArtifactTestResultsHandler(c *gin.Context) {
+    buildId := ginutil.ParseParamUint(c, "buildid")
+    artifactId := ginutil.ParseParamUint(c, "artifactid")
+
+    struct TestResults {
+    	Results     []TestResult `json:"results"`
+    	ArtifactID  uint `json:"artifactId"`
+    	Count       uint `json:"count"`
+    }
+    
+    testResults := TestResults{}
+	
+    m.Database.
+        Where(&TestResult{BuildID: buildId, ArtifactID: artifactId}).
+        Find(&testSummaries.Summaries)
+    
+    if len(testResults.Results) > 0 {
+    	testResults.Count = len(testResults.Results)
+    	testResults.ArtifactID = artifactId
+    } else {
+    	// dbnotfound error
+    	return
```

Comment:
Since they're asking for the test-results, doesn't it make sense to say that they weren't found?
Or is it that it doesn't count as an error because the request was successful and it's a valid state?

Reply:
In the API we've only returned 404 so far when otherwise only returning a single object. As soon as a list of objects is returned then an empty array has been returned instead of 404.

Mostly because of implementation details, but I'd say better to stay conventional and then we can fix the other endpoints behaviour later.